### PR TITLE
Phi-3

### PR DESCRIPTION
**Pull request by gugarosa**
# What does this PR do?

Integrates Phi-3 within `transformers`.

## Before submitting
- [X] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by gugarosa** [comment]
I bolded some questions that I had from the previous review. Could you please check them @ArthurZucker?

Done:
- Standardize how `Phi3LongScaledRotaryEmbedding` is validated/initialized. Uses the same approach defined by `Llama`.
- `swiglu` from `flash-attention` is required for now because we will need to ablate some alternative options.
- Uses `copies` for the unit tests in `test_modeling_phi3.py`.
- Use `# Ignore copy` on `Phi3ForCausalLM.forward`.
- Apply `.float()` on the `apply_rotary_pos_emb`.

Review:
- **`DbrxAttention` has an extra clipping on the QKV, which prevents the copy. Or is there a way to ignore certain lines?**
- **Use old `get_usable_length` - did not catch the reference, checked Llama and Mistral files, and they are using it as well.**
- **`Phi3Model` uses an extra dropout - no sure how I should proceed with the copy. Can we ignore lines?**

To be added:
- Integration tests will be added when models go public.

**Comment by CoderCowMoo** [comment]
I think documentation tests are fine, as they are only happening because they ran when phi-3-4k-instruct didn't exist. Waiting on approval for pull request

**Comment by ArthurZucker** [comment]
Hey! As I have mentioned offline I think that converting to llama format will allow anyone to use your model without any release (for any other frameworks), and we can add the new scalings ( SuScaledRotaryEmbedding,  YarnScaledRotaryEmbedding) separately! 

This would unlock the entire community! 🤗

**Comment by gugarosa** [comment]
As mentioned in Slack, other team already sent a PR with the previous `Phi3LongScaledRotaryEmbedding` structure 😭

**Comment by ArthurZucker** [comment]
But that should not impact the weights no?

**Comment by gugarosa** [comment]
No, let me see what I can do!

**Comment by gugarosa** [comment]
Fingers crossed that all tests will pass!

File: docs/source/en/model_doc/phi3.md
Comment (ArthurZucker) [inline docs/source/en/model_doc/phi3.md:33]:
```suggestion
- This model is very similar to `Llama` with the main difference of [`Phi3LongScaledRotaryEmbedding`], where it is used to extend the context of the rotary embeddings. The query, key and values are fused, and the MLP's up and gate projection layers are also fused.
```

File: docs/source/en/model_doc/phi3.md
Comment (ArthurZucker) [inline docs/source/en/model_doc/phi3.md:1]: Great!

File: src/transformers/models/phi3/configuration_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/configuration_phi3.py:89]: Are you sure `short_factor` and `long_factor` are best shared like this? We could support just numbers that we divid in the layer init?

File: src/transformers/models/phi3/configuration_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/configuration_phi3.py:136]: maybe missing the bos_token_id here!

File: src/transformers/models/phi3/configuration_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/configuration_phi3.py:171]: good!

File: src/transformers/models/phi3/modeling_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/modeling_phi3.py:100]: good as well! IMO using the `gemma` (or Llama)  rotary is better because it supports torch compile. This one is not super friendly

File: src/transformers/models/phi3/modeling_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/modeling_phi3.py:157]: let's remove one liners

File: src/transformers/models/phi3/modeling_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/modeling_phi3.py:179]: let's duplicate this in each class instead of having extra hierachy.

File: src/transformers/models/phi3/modeling_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/modeling_phi3.py:180]: you are appying rope in full precision, should not be casted back here

File: src/transformers/models/phi3/modeling_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/modeling_phi3.py:194]: let 's write each forward explicitly please and only inherit from the RotaryEmbedding

File: src/transformers/models/phi3/modeling_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/modeling_phi3.py:250]: thanks, last nit is to get rid of `y` to have up_states for example?

File: src/transformers/models/phi3/modeling_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/modeling_phi3.py:326]: anything that is always taken from the config should be taken from the config, same above. If they don't change then we should just pass the config

File: src/transformers/models/phi3/modeling_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/modeling_phi3.py:178]: this should be equivalent to scaling the positional ids no? (as long as the inv freq is recomputed)

File: src/transformers/models/phi3/modeling_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/modeling_phi3.py:333]:
```suggestion
```
removing because unused

File: tests/models/phi3/test_modeling_phi3.py
Comment (ArthurZucker) [inline tests/models/phi3/test_modeling_phi3.py:399]: 💘

**Review by ArthurZucker — COMMENTED** [review]
Overall LGTM, let's pay attention to ROPE and make it as simple and close to what our users are used to!

**Comment by fakerybakery** [comment]
> EDIT: this seems impossible as other frameworks are adapting to this format instead. We'll support the fused weights let's make the weights standardized for the good of the entire community! 🚀 🔥

So no chance we could get a Llama-converted version?

File: src/transformers/models/phi3/configuration_phi3.py
Comment (gugarosa) [reply src/transformers/models/phi3/configuration_phi3.py:89]: I don't know how these numbers were created now. Let's keep it for now and I will update them later once people tell me how they were created.

**Comment by gugarosa** [comment]
> Overall LGTM, let's pay attention to ROPE and make it as simple and close to what our users are used to!

I think RoPE should be clearer now and closer to what transformers proposes. What do you think?

For now, I would like to keep the short_factor and long_factor on the config.json because I am still waiting for some replies on how they were created.

File: src/transformers/models/phi3/modeling_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/modeling_phi3.py:100]: 💘

File: src/transformers/models/phi3/modeling_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/modeling_phi3.py:140]: same comment here, we should pass the config here
```suggestion
        config,
        device=None,
```

File: src/transformers/models/phi3/modeling_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/modeling_phi3.py:164]: we can split this in 2 lines, small nit!

File: src/transformers/models/phi3/modeling_phi3.py
Comment (ArthurZucker) [inline src/transformers/models/phi3/modeling_phi3.py:176]: let's be explicit and remove the one liner `_calc_scaling_factor` !

**Review by ArthurZucker — APPROVED** [review]
🔥 very clean, left 2 small nits about using the config instead of passing kwargs for ROPE, and no one liners. THat's it

File: src/transformers/models/phi3/modeling_phi3.py
Comment (LZHgrla) [inline src/transformers/models/phi3/modeling_phi3.py:968]: @gugarosa  Hi!

Why this flag is False?

**Comment by ArthurZucker** [comment]
Solving the conflicts: accept all changes from main, ignore current

File: src/transformers/models/phi3/modeling_phi3.py
Comment (gugarosa) [reply src/transformers/models/phi3/modeling_phi3.py:968]: Hello @LZHgrla!

We found out some issues with SDPA, probably related to https://github.com/huggingface/transformers/pull/30127. However, since it was merged, we will do another debugging round to ensure everything is working properly and enable it back.

File: src/transformers/models/phi3/modeling_phi3.py
Comment (gugarosa) [reply src/transformers/models/phi3/modeling_phi3.py:164]: Any advice on how? `make style` is always producing an extra line 🤣

**Review by ArthurZucker — APPROVED** [review]
🚀 great work

**Comment by ydshieh** [comment]
But since the CI in this PR is not green yet, you might consider to rebase on main once #30456 is merged into `main`.

Sorry about the mistake in the workflow file.

**Comment by ydshieh** [comment]
Well

=========== 100 passed, 47 skipped, 26 warnings in 391.83s (0:06:31) ===========

💯  

I ❤️  see this ! Thank you for the great contribution 🚀

**Comment by ydshieh** [comment]
The circle CI seems crazy to run my branch #30455. If it is not green, it's fine. The last commit `aeb6ae7ebefcd4360441c2df4f20c0bcd45ff6f5` is green anyway

https://app.circleci.com/pipelines/github/huggingface/transformers?branch=pull%2F30423

**Comment by ArthurZucker** [comment]
Merging! 👍🏻

**Comment by gugarosa** [comment]
Thanks for everything folks!

**Comment by ArthurZucker** [comment]
Thanks for yet another great contribution to the ecosystem !

**Comment by abuelnasr0** [comment]
what is `Su` and `Yarn` stands for in `Phi3SuScaledRotaryEmbedding` and `Phi3YarnScaledRotaryEmbedding`? 
I can see thee difference in the implementation is in the computation of the `scaling_factor` but is there any thing I can read more about this?

**Comment by ydshieh** [comment]
https://github.com/huggingface/transformers/pull/30423#issuecomment-2079289950

Hi @gugarosa It would very nice if you could take a look 🤗 .

**Pull request merged**
